gcovr/tests/test_args.py

```diff
@@ -0,0 +1,108 @@
+import sys
+import unittest
+from contextlib import contextmanager
+
+from ..__main__ import main, version_str
+
+
+try:
+    from StringIO import StringIO
+except ImportError:
+    from io import StringIO
+
+
+class CaptureObject:
+    def __init__(self, out, err, exception):
+        self.out = out
+        self.err = err
+        self.exception = exception
+
+
+class StringIOBuffered(StringIO):
+    def __init__(self):
+        self._data = None
+        StringIO.__init__(self)
+
+    def data(self):
+        self.close()
+        return self._data
+
+    def close(self):
+        try:
+            if self._data is None:
+                self.seek(0)
+                self._data = self.read()
+        finally:
+            StringIO.close(self)
+
+
+@contextmanager
+def ioscope(file):
+    try:
+        yield file
+    finally:
+        file.close()
+
+
+@contextmanager
+def capture(args):
+    newout = StringIOBuffered()
+    with ioscope(newout):
+        with ioscope(StringIOBuffered()) as newerr:
+            err, sys.stderr = sys.stderr, newerr
+            out, sys.stdout = sys.stdout, newout
+            try:
+                try:
+                    e = None
+                    main(args)
+                except BaseException as exception:
+                    e = exception
+                yield CaptureObject(sys.stdout.data(), sys.stderr.data(), e)
+            finally:
+                sys.stderr = err
+                sys.stdout = out
+
+
+class TestArgs(unittest.TestCase):
+
+    def test_version(self):
+        with capture(['--version']) as c:
+            self.assertEqual(c.err, '')
+            self.assertTrue(c.out.startswith('gcovr %s' % version_str()))
+            self.assertEqual(c.exception.code, 0)
+
+    def test_help(self):
+        with capture(['-h']) as c:
+            self.assertEqual(c.err, '')
+            self.assertTrue(c.out.startswith('Usage: gcovr [options]'))
+            self.assertEqual(c.exception.code, 0)
+
+    def test_empty_root(self):
+        with capture(['-r', '']) as c:
+            self.assertEqual(c.out, '')
+            self.assertTrue(c.err.startswith('(ERROR) empty --root option.'))
+            self.assertEqual(c.exception.code, 1)
+
+    def test_empty_objdir(self):
+        with capture(['--object-directory', '']) as c:
+            self.assertEqual(c.out, '')
+            self.assertTrue(c.err.startswith('(ERROR) empty --object-directory option.'))
+            self.assertEqual(c.exception.code, 1)
+
+    def test_branch_threshold_nan(self):
+        with capture(['--fail-under-branch', 'nan']) as c:
+            self.assertEqual(c.out, '')
+            self.assertTrue('not in range [0.0, 100.0]' in c.err)
+            self.assertNotEqual(c.exception.code, 0)
```

Comment:
In some places we have `assertNotEqual(code, 0)` and others `assertEqual(code, 1)`. That tripped me up because I overlooked the "not". Can we use the `assertNotEqual(code, 0)` everywhere where a specific exit code is not necessary?

Reply:
I used `assertEqual(code, 1)` when `gcovr` calls `sys.exit(1)` and `assertNotEqual(code, 0)` when parsing fails (we do not know what optparse will call sys.exit with)